Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py

```diff
@@ -397,67 +346,110 @@ def _foregroundWidths(self):
+        if not self.getProperty(Prop.LINE_POSITION_INPUT).isDefault and linePosition == 0.0:
+            linePosition = self.getProperty(Prop.LINE_POSITION_INPUT).value
+        else:
+            # Fit peak position
+            peakWSName = self._names.withSuffix('peak')
+            args = {'InputWorkspace': ws, 'OutputWorkspace': peakWSName, 'EnableLogging': self._subalgLogging}
+            if not self.getProperty(Prop.XMIN).isDefault:
+                xmin = self.getProperty(Prop.XMIN).value
+                args['RangeLower'] = common.inTOF(xmin, l1, l2, theta)
+            if not self.getProperty(Prop.XMAX).isDefault:
+                xmax = self.getProperty(Prop.XMAX).value
+                args['RangeUpper'] = common.inTOF(xmax, l1, l2, theta)
+            if not self.getProperty(Prop.START_WS_INDEX).isDefault:
+                args['StartWorkspaceIndex'] = self.getProperty(Prop.START_WS_INDEX).value
+            if not self.getProperty(Prop.END_WS_INDEX).isDefault:
+                args['EndWorkspaceIndex'] = self.getPropertyValue(Prop.END_WS_INDEX).value
+            FindReflectometryLines(**args)
+            linePosition = mtd[peakWSName].readY(0)[0]
+            self._cleanup.cleanup(peakWSName)
+        return ws, linePosition
+
+    def _addSampleLogInfo(self, ws, linePosition):
+        """Add foreground indices and linePosition to the sample logs, names start with reduction."""
+        # Add the fractional workspace index of the beam position to the sample logs of ws.
+        run = ws.run()
+        run.addProperty(common.SampleLogs.LINE_POSITION, float(linePosition), True)
+        # Add foreground start and end workspace indices to the sample logs of ws.
+        hws = self._foregroundWidths()
+        beamPosIndex = int(numpy.rint(linePosition))
+        if beamPosIndex > 255:
+            beamPosIndex = 255
+            self.self.log().warning('Is it a monitor spectrum?')
+        sign = self._workspaceIndexDirection(ws)
+        startIndex = beamPosIndex - sign * hws[0]
+        endIndex = beamPosIndex + sign * hws[1]
+        if startIndex > endIndex:
+            endIndex, startIndex = startIndex, endIndex
+        run.addProperty(common.SampleLogs.FOREGROUND_START, int(startIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_CENTRE, int(beamPosIndex), True)
+        run.addProperty(common.SampleLogs.FOREGROUND_END, int(endIndex), True)
+        # Add two theta to the sample logs of ws.
+        if not self.getProperty(Prop.TWO_THETA).isDefault:
+            twoTheta = self.getProperty(Prop.TWO_THETA).value
+        else:
+            spectrum_info = ws.spectrumInfo()
+            twoTheta = numpy.rad2deg(spectrum_info.twoTheta(int(numpy.rint(linePosition))))
+        run.addProperty(common.SampleLogs.TWO_THETA, float(twoTheta), 'degree', True)
+        return twoTheta
+
+    def _moveDetector(self, ws, twoTheta, linePosition):
```

Comment:
`twoTheta` could be read from the sample logs. It is not necessary to pass it as a value here.